modules/plugin/jdbc/jdbc-sqlserver/src/main/java/org/geotools/data/sqlserver/SQLServerDialect.java

```diff
@@ -1017,4 +1018,52 @@ public boolean canGroupOnGeometry() {
         // The type "geometry" is not comparable. It cannot be used in the GROUP BY clause.
         return false;
     }
+
+    @Override
+    public List<ReferencedEnvelope> getOptimizedBounds(
```

Comment:
Hmm... not sure this is the right place, but maybe I'm misunderstanding.

A bit of context, [here is the code using optimized bounds in JDBCDataStore](https://github.com/geotools/geotools/blob/4ef74f630509d029c09b85af2d585f8f9acf8453/modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java#L1271):

* Calls onto getOptimizedBounds
* If that returns nothing, calls onto Envelope aggregation
* In case an envelope aggregation was not available, the code is also looping over the results and accumulating bounds in memory

At the same time, the Javadoc for ``getOptimizedBounds`` states:

```
     * Returns the bounds of all geometry columns in the layer using any approach that proves to be
     * faster than the plain bounds aggregation (e.g., better than the "plain select extent(geom)
     * from table" on PostGIS), or null if none exists or the fast method has not been enabled
     * (e.g., if the fast method is just an estimate of the bounds you probably want the user to
     * enable it manually)
```

What I'm seeing here, and maybe I'm misunderstanding, is simple aggregation, which should be going in ``encodeGeometryEnvelope`` instead, while the current code in the SQL Server dialect is currently just returning the envelope of every single geometry, basically using the fall back path in JDCBDataStore.

I'm also guessing that ``EnvelopeAggregate`` was not available in the SQLServer versions we were targeting back when the store was published, but I don't remember having a list of supported SQL Server versions either so... it's probably ok.

Reply:
`EnvelopeAggregate` is available in (Microsoft) supported versions (2017 and up) I'm  not sure  about 2008 (the is a [hint](https://social.technet.microsoft.com/wiki/contents/articles/4136.new-spatial-features-in-sql-server-code-named-denali.aspx) it is, but that is no longer a supported product.

As far as I know `EnvelopeAggregate` provides an optimized, database-side aggregation of a set of geometries (in this case all of a column of a table); for ref. https://learn.microsoft.com/en-us/sql/t-sql/spatial-geometry/envelopeaggregate-geometry-data-type?view=sql-server-linux-ver15 
This should IMO be that fastest way to get the bounds of a SQL Server featuretype (but it would depend on unknown database internals such as spatial index use) 

To avoid parsing the SQL Server geometry and not parse WKT (when using toString()) I ask for the X and Y of opposing corners (points 1 and 3) of the envelope/geometry (we already know the SRID):

```
SELECT
    geometry::EnvelopeAggregate(myGeomCol).STPointN(1).STX,
    geometry::EnvelopeAggregate(myGeomCol.STPointN(1).STY,
    geometry::EnvelopeAggregate(myGeomCol.STPointN(3).STX,
    geometry::EnvelopeAggregate(myGeomCol.STPointN(3).STY
FROM myTable;
```

I could have done: `SELECT geometry::EnvelopeAggregate(myGeomCol) FROM myTable;`  and then returned  `#decodeGeometryEnvelope(ResultSet rs, int column, Connection cx)` 


I've been wondering about the javadoc of `getOptimizedBounds` as all of the implementations seem to return a list with 1 envelope (not an envelope for each geometry column of a table; just the default geometry column) and that is what this new code does, a list with 1 ReferencedEnvelope that is the envelope of the (default) geometry column of the table.

perhaps the javadoc should instead be something like:
```
     * Returns the bounds of all geometries of a column in the layer using any approach ...
```

Not sure why `encodeGeometryEnvelope` should be part of this game... it does not execute a query

Reply:
So... I could (should) execute the same SQL in a loop for all other geometry columns of the featuretype to make it match the javadoc?...